Add script to supervise initctl

kevindew opened the pull request:
Trello: https://trello.com/c/cwsh79vA/200-rollout-of-continuous-deployment

This adds a script to wrap around the calls made to "initctl start $app",
"initctl restart $app" and "initctl reload $app" as part of an
application deployment. This script adds the ability to monitor that
these calls result in a new process that is actually running and the old
process ending.

The motivation to add this is to support continuous deployment on
GOV.UK. In particular the need to know that a service is successfully
running after a restart [1]. I felt the most consistent way this could
be solved was by applying a wrapper command rather than modifying the
deployment scripts for services specifically. This wrapper is also
helpful for the other major GOV.UK app services as it will fail a
deployment if these fail to start which will mean we know of a problem
sooner.

I expect this to be applied by changing [2] from:

```
if fetch(:perform_hard_restart, false)
  run "sudo initctl start #{application} 2>/dev/null || sudo initctl restart #{application}"
else
  run "sudo initctl start #{application} 2>/dev/null || sudo govuk_unicorn_reload #{application}"
end
```
to:

```
if fetch(:perform_hard_restart, false)
  run "sudo govuk_supervised_initctl start #{application} || sudo govuk_supervised_initctl restart #{application}"
else
  run "sudo govuk_supervised_initctl start #{application} || sudo govuk_supervised_initctl reload #{application}"
end
```

I'm pretty convinced that this wrapper script is rather a nasty way to solve
these problems and I expect that it's probably resolved with better/newer
tools. It seems strange that upstart doesn't realise it started something
that failed instantly - maybe systemd knows how to deal with that? I hope
and expect this script to only live as long as we continue using Trusty.

Note this script supersedes govuk_unicorn_reload and once things are
migrated to this I will remove that.

[1]: https://github.com/alphagov/govuk-rfcs/blame/master/rfc-128-continuous-deployment.md#L143
[2]: https://github.com/alphagov/govuk-app-deployment/blob/e3f0bb339b60021fb0663aa9868cede064e8c1cd/recipes/ruby.rb#L20-L27

benthorner reviewed (commented):
Looks good! I like how the level of checking and logging is broadly consistent between the commands. I'm conscious this adds a layer of complexity, so I've got a few comments to tease out the value vs. readability compromise, so it's clearer what the scope of this script should be.

Before we start applying this:

- Have you checked this works OK with a web process and a service process, for each command? (I'm guessing you have.)
- Do you anticipate any issues for services with the assumption about the pidfile location? (I can't think of any.)

kevindew commented on modules/govuk/files/usr/local/bin/govuk_supervised_initctl:
```diff
@@ -0,0 +1,137 @@
+command=$1
+service=$2
+pid_file=/var/run/$service/app.pid
+original_pid="" && [ -e $pid_file ] && original_pid=$(cat $pid_file)
```
Ah yes, I can see that. I think I got carried away wanting as close as I could get to a ternary operator (which was not very close)

kevindew commented on modules/govuk/files/usr/local/bin/govuk_supervised_initctl:
```diff
@@ -0,0 +1,137 @@
+        status "${service} is running"
+        exit 0
+      else
+        status "${service} is an inconsistent state, stopping before starting..."
```
Yeah I added this in after some testing when I fell into a situation where it could fail to start/restart without manual intervention. Given, if you have a situation with `sudo govuk_supervised_initctl start #{application} || sudo govuk_supervised_initctl reload #{application}` and the upstart thinks the process is running but the process actually isn't (I found it quite easy to replicate this by putting in a syntax error in a Gemfile). Then you can end up with this code knowing the process isn't running but being upstart not responding to `initctl start` because it think it's running - and the reload command would also fail.

benthorner replied:
That's odd. I assumed that, while it may initially (and correctly) think the service is up (before it fails), it sounds like there's also a lag i.e. the status of according to `initctl` is incorrect. Could we also have another case, then, where `initctl` thinks the process is stopped, but it's actually running?

kevindew replied:
Yeah, I imagine you're right about that second state of running but upstart thinking otherwise. I guess I actually had that covered until I fell into fixing this bug and introduced that scenario. 

I'll have a go at replicating that later on today to see what happens - I'll also get a little more info to explain what happens in the inverse scenario (upstart thinks it's running when it's not).

kevindew replied:
I did some further digging into this. I think the oddness/complexity in this is due to there being two (main) ways of running apps, either directly as per email-alert-service or indirectly through unicornherder.

In the direct scenario things behave as you assumed and the pid file refers to the same pid that upstart is aware of. In the indirect scenario things behave in the annoying way where upstart is happily satisfied that unicornherder is running, whereas the actual app is never started. The pid file for that is the unicorn master instance rather than unicornherder, so whilst upstart is happy the service is running there is no file in /var/run/email-alert-api/app.pid.

To try illustrate with some scenarios:

## Email Alert Service with a typo in the Gemfile

- Service will have a start attempted a few times before quitting
- Upstart won't think the service is running
- There will be a pid file left in /var/run/email-alert-service/app.pid pointing to a non-existent process

## Email Alert API with a typo in the Gemfile

- Upstart thinks the service is running (because unicornherder is)
- There is no pid file created at /var/run/email-alert-api/app.pid
- You can't use `sudo govuk_supervised_initctl restart email-alert-api` because there isn't an old app process running to restart
- You can't run `sudo initctl start email-alert-api` because upstart thinks it's running, therefore needs a `sudo initctl stop email-alert-api` before it can be re-attempted. 

Hope that helps make sense of it

----

You're right that it is possible to have the inverse scenario where a process can be running and Upstart does not have knowledge of it. This scenario seems much less likely to fall into than the previous one (I couldn't think of a problem with an individual app that could cause it for instance), so I've gone with putting that as an error scenario rather than attempting to fix it.

I've got some more documentation to update off the back of this change, but let me know if this all seems sound first.

benthorner replied:
Yes, that does make sense. I checked to see `initctl` refers to the herder PID, yet the `app.pid` file refers to the child / master process, as you described. Conversely, the plain email-alert-service is consistent between the two.

(Apparently initctl does some [auto-cleverness](http://spootnik.org/entries/2014/11/09/pid-tracking-in-modern-init-systems/) to determine the PID a process is running with, which naturally doesn't work with Sidekiq workers because our initctl scripts are actually just helper loops.)

Based on this, I also agree the inverse scenario is unlikely, as it's unrelated to an issue with the application. Let me know when you're done with more documentation and I'll do a final review.

kevindew replied:
Oh cool, that link was an interesting read. It feels like, if you were prepared to go on a big yak shave, you could probably do all the checks of this script through upstart scripts.

A somewhat annoying thing about our Sidekiq upstart scripts is that nothing [actually uses the child process stuff that is part of the loop](https://github.com/alphagov/govuk-puppet/blob/b0bd7e2acbecfc51e12d0c7a547b7f8af03ca844/modules/govuk/templates/procfile-worker.conf.erb#L18-L28) (every app has a worker list of 1) so with a bit of will they could be simplified away.

I've now updated the documentation for this so it should be ready for another review.

benthorner replied:
Thanks, the latest version makes sense to me. I think composition is the right approach here: we only need the extra complexity for the deploy scripts; keeping the individual init scripts simple reduces the amount of code to understand.

After reading through again, I have a couple of minor suggestions as someone trying to understand the purpose of this script in future. One is in a separate comment. The other relates to the consistency between `start` vs `reload` / `restart`: it would be more consistent to check the pidfile first in `start`, like we do in the others.

kevindew commented on modules/govuk/files/usr/local/bin/govuk_supervised_initctl:
```diff
@@ -0,0 +1,137 @@
+case $command in
+  start)
+    status "Checking if ${service} is running"
+    if [[ $(initctl status $service) =~ "running" ]]; then
```
It's part of this condition: https://github.com/alphagov/govuk-puppet/pull/10952/files#r577065140

benthorner replied:
Minor: although it's written in the PR comments, I think it would be helpful for someone in future to explicitly state the scenario, as it's so unintuitive e.g.

```
# Checking `initctl status` doesn't necessarily mean a GOV.UK service is
# running. For example, unicornherder processes have two PIDs: one for 
# the parent "herder" process and one for the child "master" process; the 
# former is tracked by Upstart, the latter by the pidfile. If Upstart believes 
# the process is running it will need to be stopped before another attempt 
# to start the service. This is a convenience to avoid manual intervention.
```

kevindew commented on modules/govuk/files/usr/local/bin/govuk_supervised_initctl:
```diff
@@ -0,0 +1,137 @@
+
+    initctl start $service
+
+    sleep 10 # modest amount of time to allow the process to start
```
Thanks

kevindew commented on modules/govuk/files/usr/local/bin/govuk_supervised_initctl:
```diff
@@ -0,0 +1,137 @@
+    fi
+    ;;
+  restart)
+    ([ ! -z $original_pid ] && pid_running $original_pid) || error "${service} is not running"
```
Yup, I can add a bit to the commit message.

My logic was that a process wasn't restarted unless the new one had started AND the old had stopped. It also offers a bit more symmetry with reload too which is good for predictable behaviour.

benthorner replied:
Thanks. Minor: the PR message is now out-of-sync with the commit.

kevindew commented on modules/govuk/files/usr/local/bin/govuk_supervised_initctl:
```diff
@@ -0,0 +1,137 @@
+      waiting_for=$(($waiting_for+10))
+      current_pid=$(cat $pid_file) || error "Couldn't establish the current pid"
+
+      if [ -z $new_pid ] && [ $original_pid -ne $current_pid ]; then
```
Yeah - I think there's a few situations that can occur around this if there's other upstart usages at the same time. If I remember right, most of these should occur with the deploy failing which seems a preferable outcome than claiming a failed deploy was successful.

kevindew commented on modules/govuk/files/usr/local/bin/govuk_supervised_initctl:
```diff
@@ -0,0 +1,137 @@
+      if ! pid_running $original_pid; then
+        status "Process ${original_pid} has exited"
+        break
+      elif [ $waiting_for -gt 180 ]; then
```
That's a good point - I'll add that.

kevindew commented on modules/govuk/files/usr/local/bin/govuk_supervised_initctl:
```diff
@@ -0,0 +1,137 @@
+
+# This script is a wrapper around upstart initctl start/restart/reload
+# commands and is intended for applications/service that have their main
+# process in a pid file at /var/run/$sevice/app.pid - which is the case for
```
Thanks

kevindew commented:
Thanks for the review @benthorner - I've addressed all your comments.

> Have you checked this works OK with a web process and a service process, for each command? (I'm guessing you have.)

Yup I've had success trying with both. I've not tried it out as part of a deploy though, only as isolated commands. I'll try that out before I merge.

> Do you anticipate any issues for services with the assumption about the pidfile location? (I can't think of any.)

I've not found any evidence of divergence here. I think they go all through https://github.com/alphagov/govuk-puppet/blob/ee7aee7b89a8ee9b739dbb79018e0ede225392c4/modules/govuk/files/usr/local/bin/govuk_spinup which forces an approach. I figured it'd not be too hard to add an argument for the pidfile location, but couldn't see a reason why it was necessary.

One thing to note is that this doesn't work with sidekiq workers as we don't have a pid file for them (there seems to be other strange things regarding them too which keep it being not quite simple)

benthorner reviewed (approved):
Thanks for your patience with the review 🥇. It's been quite a long one, but then this does affect most deployments, so hopefully worth the extra effort.

In writing one of my closing comments, I did have a tangential thought: why can't unicornherder just quit when it's lost track of its master process? That would simplify the `start` behaviour, as the tracked PID would eventually be in-sync with the state of the master process. On the other hand, the unintuitive way Upstart divines PIDs to track is for me its own justification of checking something more concrete to corroborate.

kevindew commented:
Thanks @benthorner, I've made those suggestions and I'm doing some more trying out of this on integration.

> In writing one of my closing comments, I did have a tangential thought: why can't unicornherder just quit when it's lost track of its master process? That would simplify the start behaviour, as the tracked PID would eventually be in-sync with the state of the master process. On the other hand, the unintuitive way Upstart divines PIDs to track is for me its own justification of checking something more concrete to corroborate.

Yes, I've had similar thoughts that most aspects of this are likely bodge solutions to things that are either solved nicer with newer tools or, more simply, by better usage of the current tools.

On your specific question. It actually does [quit when it fails to create a master worker](https://github.com/gds-operations/unicornherder/blob/a89e01278490ad73a4ede89eb2acc57b19269601/unicornherder/herder.py#L220-L221). However it takes a relatively long time to get into this state (~30s), and once it's killed upstart will then start a new (so upstart is then managing a new pid).

I feel like the problem is that there is a long daisy chain of binaries involved that operate asynchronously: start-stop-daemon (which in newer versions has a --notify-await argument), to unicornherder, then unicorn and finally the unicorn worker. It feels like if you were prepared to just rethink the tooling with a requirement that you don't want anything to start until it's actually started you could get something much simpler than this with no need for any extra scripts.  Even if upstart alone was in charge of managing you might be able to do things much simpler with a post-start block of code, but as I understand systemd is better still for this.

benthorner commented:
Thanks for looking into those extra comments and rearranging the `start` function ⭐. It (still) looks good to me. 

> It actually does quit when it fails to create a master worker.

Aha, I missed that in my quick skim of the code. But I also see I neglected to consider that Upstart will keep trying to start the process when it fails, so it won't actually get in-sync - it'll just keep failing every 30s. Thanks for explaining. 

> in newer versions has a --notify-await argument

Looks cool. But of course we're not on the version with useful things like that in it 🙈.

Hopefully replatforming will do away with the herder altogether - I struggle to see why we need it if we have stuff in containers, as we should be able to gracefully transition from old to new. But who knows when we'll get there.

In the meantime, this is a handy, net improvement 🚀.

kevindew commented:
Thanks, and yup herder is definitely on the way out with replatforming world, whether we manage the switch to puma is less clear - but won't be as hard as it is now.

I'd be very sad if this script ended up being ported to the new world 😆
